types/mithril/index.d.ts

```diff
@@ -208,7 +208,7 @@ declare namespace Mithril {
 	 * Any Javascript object that has a view method can be used as a Mithril component.
 	 * Components can be consumed via the m() utility.
 	 */
-	interface Component<Attrs, State extends Lifecycle<Attrs, State>> extends Lifecycle<Attrs, State> {
```

Comment:
@sandersn 

Here's a few examples:

```ts
// Object component
interface State { count: number }
const Comp = <Component<any, State>> {
  oninit() { this.count = 0 },
  inc() { this.count++ },
  view(vnode) {
    return [
      m("div", ["Count: ", vnode.state.count]),
      m("button", {onclick: () => this.inc()}, ["Increment"])
    ]
  },
}

// Class component
class Comp implements Component<any, State> {
  count = 0
  inc() { this.count++ }
  view(vnode) {
    return [
      m("div", ["Count: ", vnode.state.count]),
      m("button", {onclick: () => this.inc()}, ["Increment"])
    ]
  }
}

// Closure component
function Comp(): Component<any, State> {
  let count = 0
  return <Component<any, {}>> {
    // In practice, `inc` would be a function instead
    inc() { count++ }
    view(vnode) {
      return [
        m("div", ["Count: ", vnode.state.count]),
        m("button", {onclick: () => this.inc()}, ["Increment"])
      ]
    }
  }
}
```

I suspect `ThisType<T>` + some intersection types *should* work, but with some added boilerplate.

@spacejack Could you chime in on this? You've been the one mostly keeping up with the definitions, so your input on this would be very valuable. 😉